### ISSUE TITLE
Reading DataSet from Streams

### ISSUE THREAD

**Issue by otto-dev** [deeplearning4j/deeplearning4j#8578]
#### Issue Description

This is a feature request to provide an inteface for steaming of DataSets. The purpose of this is to allow training and evaluation of neural networks using non-local data sources.

I've had a look for this functionality and other issues, but the only thing I found was this gitter comment from March:

> Alex Black @AlexDBlack Mar 25 17:25
> ok, so let's go back to the "read from streams" option
maybe open an issue describing what you want to do, and we can look at adding it... it's a reasonable (albeit slightly uncommon) use case that I'd like to support

To illustrate, my particular use case is that I'm using [Akka Streams](https://doc.akka.io/docs/akka/current/stream/index.html) to generate large data sets on the fly. This means that a `source` will emit pairs of `features` and `labels` as they are generated. The format of the `features` and `labels` is under my control, and could be something like a one-dimensional array of `doubles` for each row.

In principle, the desired functionality can be represented like this (pseudo-codish):

```
sourceOfRows.runForeach(
    (featureArray, labelArray) -> trainExistingNetwork(featureArray, labelArray)
)
```

My previous attempt is to use a `ListDataSetIterator` for each batch of training data (i.e. arrays of rows). I'm not sure if there are issues with this approach. If this approach is fine, then this issue might not be needed.

The example code snippet is in Scala, so I apologize if I'm speaking a foreign language here. Also, the code is a bit dated, and things may have changed slightly:

```scala
  override def trainOnBatch(inputRows: Seq[IndexedSeq[Double]], labelRows: Seq[IndexedSeq[Double]]): Unit = {
    val dataSet = new DataSet(
      toMatrix(inputRows),
      toMatrix(labelRows),
    ).asList
    network.fit(new ListDataSetIterator(dataSet, 64))
  }

  // ...

  private def toMatrix(rows: Seq[IndexedSeq[Double]]): INDArray = Nd4j.create(rows.map(_.toArray).toArray)

**Comment by saudet** [comment]
Please send a pull request, even just as a WIP. It will be easier to review! Thanks

**Comment by otto-dev** [comment]
> Please send a pull request, even just as a WIP. It will be easier to review! Thanks

If you can, could you please give me a starting point? I'll have to see if I am able to, within reasonable time. The reason why I'm opening this issue in the first place is because I'm not sure how to do this even as an end-user, and I have no familiarity with the code-base apart from the publicly visible API.

**Comment by saudet** [comment]
There's a nice guide about pull requests here:
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request
If you're talking about the code of Deeplearning4j itself, I'm afraid you'll need to read it a bit to get familiar with it, but we're here to help if you have any questions.

**Comment by AlexDBlack** [comment]
As a basic approach, the code you have provided is fine.
It might be possible to optimize it slightly - for example if your Seq is already in batches of the correct size, you can simply do:
```
  override def trainOnBatch(inputRows: Seq[IndexedSeq[Double]], labelRows: Seq[IndexedSeq[Double]]): Unit = {
    val dataSet = new DataSet(
      toMatrix(inputRows),
      toMatrix(labelRows),
    )
    network.fit(dataSet)
  }
```

If this is all you need, I think we can close this issue, and don't need to modify the API.
But if this isn't suitable for some reason, can you elaborate on why?

The only issue I can forsee here is that the INDArray (produced by your toMatrix) are allocated outside of a workspace, which means they have to be garbage collected to free off-heap memory, with a minor memory and performance hit vs. doing it in a DataSetIterator. Unless those arrays are really huge, this shouldn't be a problem in practice though.

**Comment by otto-dev** [comment]
Thank you very much @AlexDBlack, I appreciate that you had a look at this. I'll proceed with the above method, and we can close this issue - in the end I only needed  confirmation from someone more knowledgeable with dl4j.
